### ISSUE TITLE
Resolving scope type of static method call

### ISSUE THREAD

**Issue by jjbr** [javaparser/javaparser#1439]
Say I have class A with static method foo(). In class B in call foo with: A.foo();.
When I now try to resolve the type of the expression's scope (i.e. A), the SymbolSolver throws an error along the lines of: 

> Exception in thread "main" UnsolvedSymbolException{context='A', name='Solving A'}

Now resolving the type of A is odd, of course, as A is a type itself and not a variable. Does JavaParser / the SymbolSolver provide a means to easily differentiate between an instance and a static method call (e.g. myInstance.foo() and MyClass.foo())? In the latter case, is there an easy method to retrieve the type represented by the scope?

**Comment by ftomassetti** [comment]
No, it does not provide a way to differentiate. If you try to resolve the whole method call JavaParser will try to solve both as an instance or static method and figure out the answer for you. Short answer: do not try to solve the scope, solve the method call and let JavaParser figure things out for you

**Comment by malteskoruppa** [comment]
Thanks for the clarification @ftomassetti.

Would it be possible to provide a method MethodCallExpr.isStaticCall() or similar that would use the same logic behind the scenes to determine whether the call expression is an instance or a static method call as JavaParser already does when resolving the type of the method call expression?

I think it would be a great feature. :smiley: It would enable users of JavaParser to do some cool and basic things like "count the number of static method calls vs. instance method calls in this application". Currently, that wouldn't be possible, if I understood correctly?

**Comment by ftomassetti** [comment]
Currently it would not be *directly* possible, but you could simple call `resolve` and then on the `ResolvedMethodDeclaration` you could check if the declaration is static or not.

**Comment by ftomassetti** [comment]
It could be a useful feature and it sounds perfect for a Pull Request.... Just saying...

**Comment by malteskoruppa** [comment]
> Currently it would not be *directly* possible, but you could simple call `resolve` and then on the `ResolvedMethodDeclaration` you could check if the declaration is static or not.

I thought the same thing, however, it is not perfect: What about instance calls of static methods? (I mean static methods accessed via instance variables.) Such calls would be classified as static calls by this approach, which would be arguably incorrect (or wouldn't it?).

I realize that accessing static methods via instance variables is bad practice, of course. In fact, I think that it is a design flaw in Java that you are even *allowed* to do it. I can't see any reason why this would be useful, ever. There are plenty of other languages that don't allow it. It only makes for confusing code. Nevertheless, Java does allow it, and if we were to build such a feature into JavaParser, wouldn't it have to be designed in such a way that it would correctly handle such corner cases?

**Comment by ftomassetti** [comment]
Well, no, they would be classified as static calls because the MethodDeclaration is marked as static

**Comment by malteskoruppa** [comment]
Exactly, that's what I mean. And wouldn't this classification be incorrect? I'm not even sure there. :wink: But I tend to think that a method call expression of a static method via an instance variable is still an *instance call* from the caller's point of view. So to be completely accurate, `MethodCallExpr#isStatic()` would have to return `false` in this case (even though the method definition is static). And that's non-trivial to determine, isn't it?

Admittedly, this is a corner case that shouldn't ever happen when parsing "good" Java code that doesn't do such confusing things. :smile:

**Comment by ftomassetti** [comment]
Well, I disagree because to me it would be a static call because that call would not have access to any instance. It would just be a weird usage of the syntax. But I can see that we may want a way to determine if the scope is a value or a type.

**Comment by malteskoruppa** [comment]
> Well, I disagree because to me it would be a static call because that call would not have access to any instance.

Yes, that's true; I see that too, which is why I said it would be "arguably" incorrect to classify such a call as static. But I guess you're right, we would be quite justified in calling such a call "static", even though the scope is a value.

> But I can see that we may want a way to determine if the scope is a value or a type.

I agree, it would be really cool if there was a way to determine this.

**Comment by malteskoruppa** [comment]
Concerning that PR for a method `MethodCallExpr.isStaticCall()` that we talked about...

After our discussion I'm not even sure that this would be useful anymore. :wink: 

Wouldn't the implementation of that method essentially look like

```
public final class MethodCallExpr extends ... implements ... {
//...
  public boolean isStaticCall() {
    return resolveInvokedMethod().isStatic();
  }
//...
}
```

This is so easy that I'm wondering whether it actually yields any benefit? It also feels a bit out of place. If `MethodCallExpr` has such a method, then `FieldAccessExpr` could also get a method `isStaticAccess()` or so to determine whether the accessed field is static. But there too, I assume that one could do the very same thing by getting a `ResolvedFieldDeclaration` from the `FieldAccessExpr` somehow (not entirely sure how :wink:).

**Comment by matozoid** [comment]
@ftomassetti @malteskoruppa - will anything more happen in this issue? :-)

**Comment by ftomassetti** [comment]
Not sure...

**Comment by malteskoruppa** [comment]
Not sure either. :smiley: 

Let me sum this up: JavaParser's symbol solver can be used to determine, for a method call expression, whether the called method is static (`resolveInvokedMethod().isStatic()`). However, it cannot be used to determine whether the scope of a `MethodCallExpr` is a type (e.g., `SomeClass.foo()`) or an expression that evaluates to a variable (e.g., `someObj.foo()`). In fact, the symbol solver does somehow figure it out internally (it has to, since it has to find the called method), but it is not directly/easily possible to make this information available to the JavaParser user.

While using `resolveInvokedMethod().isStatic()` to determine if the method is static may be enough in many scenarios, there are scenarios where it is not. As a (maybe somewhat contrived) example, let's say someone would want to build a compiler for Java source code on top of JavaParser. In that scenario, one would have to know whether to evaluate the method call expression's scope (in case it is an expression that resolves to a variable) or not to evaluate it (in case it is a type).

I would love to see the possibility to make that case distinction in JavaParser. However, I had a look at `resolveInvokedMethod()` and followed its logic across the methods it calls a little bit, and I would have to agree with @ftomassetti that it seems that this case distinction is not easily implemented: I couldn't even figure out at what point the symbol solver figures it out internally. (Somehow it doesn't really seem to *explicitly* figure it out.)
